javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
**IMPORTAT UPDATE**

When I interprete this right

> anything that was visible to thread A when it writes to volatile field f becomes visible to thread B when it reads f.

then the check `value == UNDEFINED` may not reflect the actual main memory state.

**Example:**

1. Thread A calls get(), acquires the lock and writes `value` and `supplier` (but is still in the synchronized block)
2. Thread B calls get() and checks `value == UNDEFINED`. Because Thread B did not read `supplier`, the initialized `value` gets not visible to Thread B. Then `supplier != null` is checked, which is `false`because Thread A set it to null. Because supplier was read, the initialized `value` is now visible to Thread B and the `return value` makes sense.

If the above cannot be proved to be wrong then, for the sake of correctness and safety, we should remove the first `value == UNDEFINED` check completely again. What remains is this:

```java
    private static final Object UNDEFINED = new Object();

    private transient volatile Supplier<? extends T> supplier;

    @SuppressWarnings("unchecked")
    private T value = (T) UNDEFINED;

    public T get() {
        if (this.supplier != null) {
            synchronized(this) {
                final Supplier<? extends T> supplier = this.supplier;
                if (supplier != null) {
                    value = supplier.get();
                    this.supplier = null;
                }
            }
        }
        return value;
    }
```

This is exactly the solution proposed in #1576.

Also `isEvaluated()` stays as is.